### ISSUE TITLE
API call from token provider causing ANR issue

### ISSUE THREAD

**Issue by JayMehta97** [GetStream/stream-chat-android#3314]
**Describe the bug**
We are using the token provider from the SDK which has a method `loadToken` that returns a string. From there we make an API call to our backend to generate a token. As the function needs to return a string, we need to make the API call using `runBlocking` using following code:

```
val streamTokenProvider = object : TokenProvider {
            override fun loadToken(): String = getChatToken()
        }
```

```
fun getChatToken(): String {
        var chatToken = ""
        runBlocking {
            when (val result = repository.getChatToken()) {
                is ResponseResource.Success -> {
                    result.value.chatTokenDetails.token?.let { token ->
                        chatToken = token
                    }
                }
                is ResponseResource.Failure -> {
                    handleTokenGenerationFailure.invoke()
                }
            }
        }
        return chatToken
    }
```

This is leading to generate ANR for some user. How can we make it asynchronous?

**SDK version**
- 4.28.2

**Expected behavior**
Stream iOS SDK provides a async callback in the token provider with which we can call this same backend API and once the API request is completed then give the new token in a completion handler, can we have something similar in Android SDK as well?

**Comment by filbabic** [comment]
Hey @JayMehta97 !

You shouldn't be using `runBlocking` in your code as that's definitely going to push an ANR as it blocks the main thread.

Can you give me a bit of info on what the `repository.getChatToken()` function provides.

The way I would go about this is for the TokenProvider to call a function which fetches the current value of the token in the `Repository`. If the value doesn't exist, then you can return an empty value for example. This value can be cached in a local property, a Flow or LiveData or a DB/shared prefs.

But you can also make sure that the token is always present by the time you set up our client and the token provider - you could use coroutines to launch a function which will fetch the token from the Repo, store it/cache it and then build the client asynchronously. That way, before the client is ready, you don't force the UI on the user that shows channels/messages, so you can do it within your splash screen lifecycle.

**Comment by kanat** [comment]
Hey @JayMehta97, @filbabic

This issue seems to be a side-effect of https://github.com/GetStream/stream-chat-android/pull/1953.
`TokenProvider.loadProvider` is forced to be invoked inside the UI thread from `setUser` method:
```kotlin
public fun setUser(
    user: User,
     tokenProvider: TokenProvider,
     listener: InitConnectionListener? = null,
) {
    // here loadToken is getting invoked in UI thread
     if (tokenUtils.getUserId(tokenProvider.loadToken()) != user.id) { 
         logger.logE("The user_id provided on the JWT token doesn't match with the current user you try to connect")
         listener?.onError(ChatError("The user_id provided on the JWT token doesn't match with the current user you try to connect"))
         return
     }
```

cc @adasiewiczr, @JcMinarro

**Comment by adasiewiczr** [comment]
hey @JayMehta97,
thanks for reporting this issue. It has been fixed in the PR mentioned above and will be released next week.
We'll notify you once the new version is available.

**Comment by JayMehta97** [comment]
Hey @adasiewiczr @kanat @filbabic,

Thanks for the updates. Looking forward to the next release.

**Comment by adasiewiczr** [comment]
hey @JayMehta97,
we just released a new version (5.1.0) with the fix mentioned above. You can find more details [here](https://github.com/GetStream/stream-chat-android/releases/tag/v5.1.0).

**Comment by JayMehta97** [comment]
Hey @kanat @adasiewiczr @filbabic, 

So with this update if I use the same code block that I had shared in the original question then it should work fine, right?
